_posts/2022-02-04-shard-indexing-backpressure-in-opensearch.markdown

```diff
@@ -0,0 +1,205 @@
+* Control knobs to run the feature either in Shadow-Mode or Enforced-Mode. Shadow-mode is a dry-run mode where only rejection breakdown metrics are published while no actual rejections are performed.
+
+
+## Step by step walk through
+
+Shard level indexing backpressure focuses on dynamically assigning memory quota from the total memory pool to each shard based on its operational needs (i.e. the incoming workloads). A combination of shard level memory utilization and throughput enables OpenSearch to decide whether it should process a request. This also prevents other nodes in the cluster from running into cascading failures due to the performance degradation caused by slow node, stuck tasks, resource intensive requests, traffic surge, skewed shard allocation, etc. Shard level indexing backpressure depends on primary and secondary parameters to make decisions.
+
+### Primary Parameters
+
+Primary parameters, are leading indicators for node duress and are governed through soft limits. We have two such primary parameters:
+
+1. **Shard memory limit breach:** The shard memory limit is breached if the allocated shard level memory limit is exceeded and there is a need to increase the allocation. At any given point the memory distribution algorithm will try to maintain that the current memory utilization of shard approximately 85% of the assigned shard memory. If the occupancy goes beyond 95%, this parameter is assumed to be breached.
+2. **Node memory limit breach:** The node memory limit is breached if the total allocated node level memory limit for indexing requests is exceeded. At any point the current memory utilization at the node level goes beyond 70% of total assigned node memory for indexing work, this parameter is assumed to be breached.
+
+Breach of primary parameters reflect no actual rejections yet, but triggers an evaluation of the secondary signals.
+
+### Secondary Parameters
+
+These parameters are based on granular, shard level performance such as request throughput and successful completion of requests. While these shard states are updated with every request, they are evaluated only when the primary parameters are breached. We have introduced two secondary parameters:
+
+1. **Throughput:** The framework tracks request latency per unit byte at the shard level and any substantial decrease in the shard throughput against its historic performance is considered a degradation.
+2. **Successful Request:** The framework also tracks successful indexing requests at the shard level and signals if there are large number of pending requests. This is used to detect memory build-ups due to stuck requests proactively and invoke the required backpressure.
+
+Indexing backpressure takes effect when one of the primary parameters and secondary parameters’ thresholds are breached on the node. This results in rejection of new incoming requests/tasks for the affected shards on the node
```

Comment:
Add a period and an another line. It is rendering as a single paragraph.


![Screen Shot 2022-02-08 at 3 45 17 PM](https://user-images.githubusercontent.com/1152927/153088281-9e68a8fc-3a19-4d6d-8584-92417a2f7d9c.png)